honeybadgermpc/betterpairing.py

```diff
@@ -394,15 +409,28 @@ def rand(seed=None):
         out = PyG2()
         if seed is None:
             seed = []
-            for _ in range(4):
+            for _ in range(8):
                 seed.append(random.SystemRandom().randint(0, 4294967295))
-            out.rand(seed[0], seed[1], seed[2], seed[3])
+            out.rand(seed)
         else:
             assert type(seed) is list
-            assert len(seed) == 4
-            out.rand(seed[0], seed[1], seed[2], seed[3])
+            assert len(seed) <= 8
+            out.rand(seed)
         return G2(out)
 
+    # length determines how many G2 values to return
+    @staticmethod
+    def hash(*args, length=1):
+        bytebuf = pickle.dumps(args)
```

Comment:
I'm wary of using pickle for this. I think there's no guarantee that pickle gives canonical encodings, there may be multiple valid encodings produced at different times or on different platforms for "equal" objects.

Reply:
Suggest making this hash explicitly require a bytes-like object as input

Reply:
I'd counter that this would make the interface more cumbersome while not necessarily reducing errors. The result I think would be requiring a pickle import on every file that calls for hashing and the call would just be G1.hash(pickle.dumps(...)) anyway. 

Regardless, there shouldn't be encoding differences for G1 and G2, as they are encoded using their affine coordinates. ZR is straightforward, so they one thing I wouldn't be completely sure about is GT, but it looks like all the values are compared directly, which would mean that an encoding difference would imply inequality.

If we're talking about serializing objects outside of the betterpairing library, then of course I can't speak for that. I'm not familiar with real world examples of how this can come up. We probably should look into something more secure than pickle anyway though.